src/app/FakeLib/TargetHelper.fs

```diff
@@ -345,13 +382,72 @@ let run targetName =
                     traceEndTarget target.Name                
         with
         | exn -> targetError targetName exn
+
+    let parallelJobs = environVarOrDefault "fake-parallel-jobs" "1" |> Int32.Parse
       
     let watch = new System.Diagnostics.Stopwatch()
     watch.Start()        
     try
         tracefn "Building project with version: %s" buildVersion
-        PrintDependencyGraph false targetName
-        runTarget targetName
+
+        if parallelJobs > 1 then
+            // multi threaded build
+
+            tracefn "Running parallel build with %d workers" parallelJobs
+
+            let queue = System.Collections.Concurrent.ConcurrentBag()
+            let finished = new System.Threading.CountdownEvent(0)
+            let cancel = new System.Threading.CancellationTokenSource()
+            let ct = cancel.Token
+
+            let order = determineBuildOrder (getTarget targetName)
+
+            let worker (threadState : obj) =
+                try
+                    while true do
+                        ct.ThrowIfCancellationRequested()
+                        match queue.TryTake() with
+                            | (true, target) -> 
+                                try
+                                    traceStartTarget target.Name target.Description (dependencyString target)
+                                    let watch = new System.Diagnostics.Stopwatch()
+                                    watch.Start()
+                                    target.Function()
+                                    addExecutedTarget target.Name watch.Elapsed
+                                    traceEndTarget target.Name       
+                                finally
+                                    finished.Signal() |> ignore
+                            | _ ->
+                                System.Threading.Thread.Sleep(100)
+                with :? System.OperationCanceledException ->
+                    ()
+
+            let workers =
+                Array.init parallelJobs (fun i -> 
+                    let thread = new System.Threading.Thread(System.Threading.ThreadStart(worker), IsBackground = true)
```

Comment:
As you use unordered `ConcurrentBag` for targets, I think it's better to use the thread pool instead of dedicated threads. Just create an `async` for each `Target` and start all of them in parallel (`Async.Parallel >> Async.RunSynchronously`). To set degree of parallelism you could use a `SemaphoreSlim`. Another advantage is that cancellation is handled automatically. As an alternative, you could use raw TPL (`seq of targets -> AsParallel -> WithDegreeOfParallelism -> ToArray`).